README
-*- text -*--------------------------------------------------
file: README
General info and compilation instructions for MASS2
-------------------------------------------------------------
-------------------------------------------------------------
Battelle Memorial Institute
Pacific Northwest Laboratory
-------------------------------------------------------------
-------------------------------------------------------------
Created November 17, 1999 by William A. Perkins
Last Change: 2014-06-08 10:57:36 d3g096
$Id$
-------------------------------------------------------------

MASS2 is now exclusively parallel.  The serial version is no longer
functional.  

-------------------------------------------------------------
Requirements
-------------------------------------------------------------

CGNS, Version 2.5, or later
---------------------------

Global Arrays, Version 5.0, or later
------------------------------------

PETSc, Version 3.0 or later
---------------------------

-------------------------------------------------------------
Example Configuration
-------------------------------------------------------------


PE10900:
set prefix = /net/flophouse/files0/perksoft/macosx

configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FCFLAGS="-g -fbounds-check -ffpe-trap=invalid,zero,overflow" \
    FC_CPPFLAG="-x f95-cpp-input" \
    PETSC_DIR="$prefix/../petsc-3.4.4" \
    PETSC_ARCH="arch-macosx-fortran-debug" \
    --enable-maintainer-mode \
    --with-cgns="$prefix/cgns2" \
    --with-global-arrays="$prefix/mass2" \
    --prefix="$prefix/mass2"

configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FCFLAGS="-O2 -g -ffpe-trap=invalid,zero,overflow" \
    FC_CPPFLAG="-x f95-cpp-input" \
    PETSC_DIR="$prefix/../petsc-3.4.4" \
    PETSC_ARCH="arch-macosx-fortran-opt" \
    --enable-maintainer-mode \
    --with-cgns="$prefix/cgns2" \
    --with-global-arrays=$prefix/mass2 \
    --prefix="$prefix/mass2"

Flophouse, with Intel compiler:

source /usr/software/intel101/x86_64/bin/ifortvars.csh
set path = (/usr/local/openmpi-intel/bin $path)
configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FC_CPPFLAG="-fpp" \
    FCFLAGS="-O2 -fpe0 -heap-arrays" \
    GA_MPILIBS="-llapack -lblas" \
    PETSC_DIR="/usr/local/openmpi-intel" \
    --with-cgns="$prefix/cgns2" \
    --with-global-arrays=/usr/local/openmpi-intel \
    --prefix="$prefix/mass2"


configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FC_CPPFLAG=-fpp \
    FCFLAGS="-g -check bounds -fpe1 -heap-arrays" \
    GA_MPILIBS="-llapack -lblas" \
    PETSC_DIR="/usr/local/openmpi-intel" \
    --with-cgns="$prefix/cgns2" \
    --with-global-arrays=/usr/local/openmpi-intel

set prefix = /net/flophouse/files0/perksoft/linux64
set path = ($prefix/openmpi44/bin $path)
configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FC_CPPFLAG=-fpp \
    FCFLAGS="-g -fbounds-check -ffpe-trap=invalid,zero,overflow" \
    PETSC_DIR="/usr/local/openmpi-intel" \
    --with-cgns="$prefix/cgns2" \
    --with-global-arrays=/usr/local/openmpi-intel
    


Grapes:

sh configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FC_CPPFLAG=-fpp \
    FCFLAGS="-O2 -fpe0 -heap-arrays" \
    GA_MPILIBS="-libverbs -llapack -lblas" \
    PETSC_DIR="/home/d3g096/stuff/src/petsc-3.1-p7" \
    PETSC_ARCH="linux-gnu-c-opt" \
    LIBS="-L$HOME/stuff/lib -lhdf5 -lsz -lz" \
    --with-global-arrays=/share/apps/ga/5.0.1/intel \
    --with-cgns=$HOME/stuff


Briareus:

module load intel/11.1
module load mpi/openmpi/1.4.3/intel

sh configure \
    FC=mpif90 \
    MPIFC=mpif90 \
    FC_CPPFLAG=-fpp \
    FCFLAGS="-O2 -fpe0 -heap-arrays" \
    GA_MPILIBS="-libverbs -llapack -lblas" \
    PETSC_DIR="/home/d3g096/stuff/petsc-3.4.4" \
    PETSC_ARCH="intel-fortran-opt" \
    --with-global-arrays=/share/apps/ga/5.0.1/intel \
    --with-cgns="$HOME/stuff/cgns2"

Franklin:

module load PrgEnv-pgi
module load ga/5.0
module load petsc/3.1_O
module load hdf5
sh configure \
    FC=ftn \
    MPIFC=ftn \
    FC_CPPFLAG="-Mpreprocess" \
    FCFLAGS="-fast" \
    LDFLAGS="-L$CRAY_HDF5_DIR/hdf5-pgi/lib" \
    LIBS="-ldl" \
    --with-global-arrays=/usr/common/usg/ga/5.0.2.i4/pgi \
    --with-cgns=$HOME/stuff/$NERSC_HOST \
    --prefix=$HOME/stuff/$NERSC_HOST


-------------------------------------------------------------
The stuff below is very dated ...
-------------------------------------------------------------

-------------------------------------------------------------
CGNS
-------------------------------------------------------------

MASS2 also requires Version 2.1 of the CGNS library.  This can be
obtained from http://www.icemcfd.com/cgns.  On Linux platforms, it is
by default compiled with gcc and g77, and the -fno-second-underscore
flag is used on g77.  Compile CGNS as follows

    SGI: make SYSTEM=mips4_n32 ==> lib/libcgns.mips4_n32.a
    Intel Linux: make SYSTEM=linux ==> lib/libcgns.linux.a
    Alpha Linux: cp Makefile.linux Makefile.alinux; make SYSTEM=alinux ==> lib/libcgns.alinux.a

The second to last line in the MASS2 configuration file controls the
CGNS output.  The following is expected to be on that line:

    1. (T/F) Flag to turn CGNS output on.  CGNS output can be on at
       the same time as NetCDF, but there is really no reason to do
       that.  

    2. (T/F) Flag to cause CGNS output to be cell-centered.  If true,
       the CGNS output will have a grid like the input grid (and
       output in gridplot1.dat) and all output values will be
       cell-centered.  Otherwise, the output grid will be like that
       produced in the NetCDF file: vertexes at model cell centers
       and artifical edges.  

    3. (T/F) Flag to add descriptions to solution output fields.  If
       true, each field of each solution has descriptors for the field
       name and the units.  This is meant to duplicate the NetCDF
       behavior.  It does not take a great deal of extra space, but
       the extra nodes can cause the model to slow down.

    4. (integer) Maximum number of time planes to put in the CGNS
       output files.  The number of time planes can be limited to
       avoid the problem that CGNS has with lots of nodes in the
       file.  

CGNS has some problems when there are a lot of nodes in the file.  Is
some tests of SAC problems, putting 1050 time planes in a single CGNS
output file doubled the MASS2 execution time (and the extra time was
not spent in I/O).  Whereas, putting a maximum of 100 time planes in
11 files increased the MASS2 execution time only 3%.  

-------------------------------------------------------------
compilation via GNU autotools
-------------------------------------------------------------

MASS2 is configured and compiled using the GNU autotools.  If you
check MASS2 out of CVS, you must build the configure script on a
system that has the GNU autotools installed (Linux systems usually
do).  The configure script is built by the command

    sh bootstrap.sh

You do not have to do this on the machine you which to compile for;
the only requirement is that autotools are installed (autoconf and
automake).  

To compile MASS2, the configure script must be run on the architecture
you wish to run MASS2.  The following architectures and compilers are
supported: 

    * Silicon Graphics with SGI Fortran 90 compiler (f90),
    * OSF4 on DEC/Compaq/HP Alpha with the DEC/Compaq/HP Fortran 90
      compiler (f90, f95), 
    * Linux on DEC/Compaq/HP Alpha with the DEC/Compaq/HP Fortran 90
      compiler (fort),
    * Linux on Intel with any of 
        * Absoft Fortran 95/90 (version 7.5 or above) compiler (f90, f95),
        * Intel Fortran (version 7.0 or above) compiler (ifc) , or
        * Lahey Fortran 95 Pro (version 6.1 or above) compiler (lf95).

If you try to configure without when not working with one of these
architectures, the configure script will crash with an error.  If
however, you set the F77 environment variable to the compiler you want
(or use --with-f90, see below) configure will try to make it work with
some (probably wrong) compilation flags.  You can also use the F77
environment variable (or --with-f90) to choose the compiler when there
are multiple choices.

The configure script has several options.  To get a complete list,
issue the command

    configure --help

The options specific to the compilation of MASS2 are as follows:

    --with-f90=cmd

        Use "cmd" as the Fortran 90 compiler command.

    --enable-f90-optimize

        Optimization is turned on using flags appropriate to the
        compiler/architecture.

    --enable-f90-debug (default)

        Debugging flags are included by default.  If
        --enable-f90-optimize is specified, the debugging flags are
        not included.  Use --enable-f90-debug to include debugging
        with optimization.

    --enable-f90-profile

        Turn on profiling.

    --with-extra-f90flags=flags

        Use the specified flags in addition to the flags added by the
        above options.  This is mostly used to add CPU specific
        optimization options.  For example, with the Intel ia32
        compiler, these flags are used to compile two different
        executables specialized for the Pentium 3 and Pentium 4:

            P3: -tpp6 -xK
            P4: -tpp7 -xW

        The same thing can be accomplished by setting F90FLAGS or
        FFLAGS in the environment before running configure.


ChangeLog
2015-02-20  William A. Perkins  <william.perkins@pnnl.gov>

	* block_initialization.f90 (initialize): initialize scalars with
	species-specific initial value

	* scalar_solve.f90 (scalar_solve): force scalar solution in dead
	cells 

	* scalars_source.f90 (scalar_source_read): add "INITIAL" keyword
	to specify initial concentration

	* plot_cgns.f90 (plot_cgns_setup, plot_cgns_write)
	(plot_cgns_file_close): close and reopen CGNS output between
	output time steps to preserve HDF-based integrity

2015-01-13  William A. Perkins  <william.perkins@pnnl.gov>

	* scalar_solve.f90 (apply_scalar_bc, scalar_solve): implement left
	bank concentration BC's

2015-01-06  William A. Perkins  <william.perkins@pnnl.gov>

	* test/transport/mixing/Makefile: make parallel

	* plot_cgns.f90 (plot_cgns_write): incorrect output of sediment
	bed mass per unit area; incorrect direction for particulate bed
	exchange 

	* scalars_source.f90 (scalar_source_post_timestep): particulate
	bed exchange variable needed putting

2014-08-14  William A. Perkins  <william.perkins@pnnl.gov>

	* test/transport/1d-analytic-1blk: make parallel and change
	picture to use gage data

	* test/transport/pointsource: make parallel and update so pictures
	can be made from CGNS output

	* test/transport/mixing: make parallel and update so pictures can
	be made from CGNS output

2014-06-30  William A. Perkins  <william.perkins@pnnl.gov>

	* bed.f90 (bed_particulate_exchange, bed_sediment_exchange): only
	report negative values when they are "bad"

2014-06-17  William A. Perkins  <william.perkins@pnnl.gov>

	* mass2_parallel.f90 (output): only print the scalar mass balance
	if there is one

2014-06-16  William A. Perkins  <william.perkins@pnnl.gov>

	* scalar_solve.f90 (transport_precalc): call to bedshear() no
	longer necessary 

	* bed.f90 (bed_initialize): remove extraneous put

	* hydro_solve.f90 (hydro): call block_hydro_update_bc() which
	contains all the stuff that used to be here

	* block_hydro_bc.f90 (block_hydro_update_bc): add routine that
	collects necessary tasks to prepare for a hydro iteration (BC,
	dead cells, etc.)

2014-06-13  William A. Perkins  <william.perkins@pnnl.gov>

	* plot_cgns.f90 (plot_cgns_write): used wrong index for collecting
	particulate 

	* bed_var.f90 (bed_var_allocate): initialize %slices

	* bed_functions.f90 (bed_post_timestep): use bed_put()

	* bed.f90 (bed_put): put pore flux for all species

	* gage_output_module.f90 (gage_print_text): particulate bed mass
	was incorrect

2014-06-10  William A. Perkins  <william.perkins@pnnl.gov>

	* scalar_solve.f90 (transport_precalc): need to compute bedshear

	* plot_cgns.f90 (plot_cgns_write): output some bed variables; more
	needed 

	* particulate_source.f90 (part_bed_exch, part_dissolve_bed_exch):
	enable the old calculations

	* mass2_parallel.f90 (update): turn bed accounting back on

	* hydro_solve.f90: add some GA after gets

	* hotstart.f90: add some GA after gets

	* block_parallel.f90 (block_module): require an index for the
	variable 

	* block_initialization.f90: add some GA sync's after gets

	* bed_functions.f90 (bed_post_timestep): add

	* Makefile.am: adjust some object/module dependencies 

2014-06-09  William A. Perkins  <william.perkins@pnnl.gov>

	* sediment_source.f90 (sediment_erosion): turn erosion on
	(sediment_source_term): deposition and erosion rates for output
	are not set here

2014-06-08  William A. Perkins  <william.perkins@pnnl.gov>

	* bed.f90 (bed_collect): added

2014-05-08  William A. Perkins  <william.perkins@pnnl.gov>

	* hydro_solve.f90 (depth_solve): initialize mass_source each call
	and set mass_source to zero in dead cells

	* block_hydro.f90 (block_mass_source_sum): only count locally
	owned mass source

	* mass_source_output.f90 (mass_file_setup, mass_print)
	(mass_file_close): only write from process zero

2014-05-07  William A. Perkins  <william.perkins@pnnl.gov>

	* transport_only_module.f90 (transport_only_check): do a better
	check of time range in transport only solutions

	* scalars_module.f90 (allocate_scalarblock_components): allocate
	srcflux 

	* scalar_solve.f90 (default_scalar_bc): initialize srcconc and
	srcflux 
	(apply_scalar_bc): new "IN SOURCE FLUX" condition recognized
	(apply_scalar_source): apply new "IN SOURCE FLUX" condition

2014-05-01  William A. Perkins  <william.perkins@pnnl.gov>

	* scalar_solve.f90 (apply_scalar_bc, apply_scalar_source): scalar
	concentration conversion was not applied for SOURCE BCs

2014-04-25  William A. Perkins  <william.perkins@pnnl.gov>

	* plot_output.f90 (plot_print): compute velocity shift before diag

	* block_hydro_bc.f90 (block_compute_bc_flux): do not compute areas

hydro_bc.f90
! ----------------------------------------------------------------
! file: hydro_bc.f90
! ----------------------------------------------------------------
! ----------------------------------------------------------------
! Battelle Memorial Institute
! Pacific Northwest Laboratory
! ----------------------------------------------------------------
! ----------------------------------------------------------------
! Created December 30, 2010 by William A. Perkins
! Last Change: 2014-04-24 13:05:55 d3g096
! ----------------------------------------------------------------

! ----------------------------------------------------------------
! MODULE hydro_bc_module
! ----------------------------------------------------------------
MODULE hydro_bc

  USE utility
  USE config
  USE table_boundary_conditions

  IMPLICIT NONE

  CHARACTER (LEN=80), PRIVATE, SAVE :: rcsid = "$Id$"

  ! ----------------------------------------------------------------
  ! TYPE bc_spec_struct
  ! ----------------------------------------------------------------
  TYPE  bc_spec_struct

     CHARACTER (LEN=10) :: bc_loc
     CHARACTER (LEN=10) :: bc_type, bc_kind, bc_extent
     INTEGER :: block, table_num != 0
     INTEGER :: con_block, con_start_cell(max_cell_values), con_end_cell(max_cell_values)
     INTEGER :: start_cell(max_cell_values), end_cell(max_cell_values)
     INTEGER :: num_cell_pairs != 0
     DOUBLE PRECISION, POINTER :: flux_area(:)

  END TYPE bc_spec_struct

  ! ----------------------------------------------------------------
  ! block_bc_struct
  ! ----------------------------------------------------------------
  TYPE block_bc_struct

     INTEGER  :: num_bc != 0
     TYPE(bc_spec_struct) :: bc_spec(200)

  END TYPE block_bc_struct

  TYPE(block_bc_struct), ALLOCATABLE :: block_bc(:)
  CHARACTER (LEN=80), PRIVATE :: bcspecs_name = "bcspecs.dat"
  INTEGER, PARAMETER, PRIVATE :: bcspec_iounit = 18

CONTAINS


  ! ----------------------------------------------------------------
  ! SUBROUTINE allocate_block_bc
  ! ----------------------------------------------------------------
  SUBROUTINE allocate_block_bc(max_blocks)
    IMPLICIT NONE
    INTEGER :: max_blocks, alloc_stat, i

    ALLOCATE(block_bc(max_blocks), STAT = alloc_stat)
    IF(alloc_stat /= 0)THEN
       CALL  error_message('allocation failed for the array of block bc ', fatal=.TRUE.)
    END IF
    CALL status_message('allocation successful for array of block bc')

    block_bc%num_bc = 0
    DO i=1,max_blocks
       block_bc(i)%bc_spec%table_num = 0
       block_bc(i)%bc_spec%num_cell_pairs = 0
    END DO

  END SUBROUTINE allocate_block_bc

  !##########################################################################
  
  SUBROUTINE read_bcspecs(max_blocks, xmax, ymax)

    ! reads the bc spec file
    ! format: block#	bc_loc	bc_type	bc_kind	bc_extent	'connect_block OR filename' 'cell pairs'			

    IMPLICIT NONE
    INTEGER, INTENT(IN) :: max_blocks, xmax(:), ymax(:)

    INTEGER :: junk1, block, num_bc, con_block, cells(2*max_cell_values) = -999 
    INTEGER :: table_count = 0, num_cells, num_cell_pairs
    CHARACTER (LEN=10) :: junk_char1, junk_char2
    CHARACTER (LEN=10) :: bc_loc, bc_type, bc_kind, bc_extent
    CHARACTER (LEN=80) :: file_name
    CHARACTER (LEN=1024) :: msg

    INTEGER :: line, lerr, ierr, maxidx

    INTEGER :: cmin, cmax, p

    CALL open_existing(bcspecs_name, bcspec_iounit)

    ! first we need to count how many TABLE bc types 

    DO WHILE(.TRUE.)

       READ(bcspec_iounit,*,END=100)junk1,junk_char1,junk_char2
       SELECT CASE (junk_char2)
       CASE ("TABLE", "SOURCE", "SINK")
          max_tables = max_tables + 1
       END SELECT
    END DO

    ! now allocate the number of table bc structs that we need
100 CALL allocate_table_bc(max_tables)

    ! now start over to fill and parse
    REWIND(bcspec_iounit)

    line = 0
    ierr = 0

    DO WHILE(.TRUE.)
       READ(bcspec_iounit,*,END=200)block, bc_loc, bc_type, bc_kind, bc_extent
       line = line + 1
       lerr = 0

       block_bc(block)%num_bc = block_bc(block)%num_bc + 1
       num_bc = block_bc(block)%num_bc

       block_bc(block)%bc_spec(num_bc)%bc_loc = bc_loc
       block_bc(block)%bc_spec(num_bc)%bc_type =	bc_type
       block_bc(block)%bc_spec(num_bc)%bc_kind =	bc_kind
       block_bc(block)%bc_spec(num_bc)%bc_extent = bc_extent
       NULLIFY(block_bc(block)%bc_spec(num_bc)%flux_area)

       ! check block number

       IF (block .GT. max_blocks .OR. block .LT. 1) THEN
          lerr = lerr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": block number out of range: ", block
          CALL error_message(msg, fatal=.FALSE.)
          ierr = ierr + lerr
          CYCLE
       END IF
       block_bc(block)%bc_spec(num_bc)%block = block

       ! make sure we understand the BC
       ! location

       SELECT CASE(bc_loc)
       CASE ("US", "DS")
          maxidx = ymax(block) - 1
       CASE ("LB", "RB")
          maxidx = xmax(block) - 1
       CASE ("IN")
          ! decide maxidx below
       CASE DEFAULT
          lerr = lerr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": ", TRIM(bc_loc), "?"
          CALL error_message(msg, fatal=.FALSE.)
       END SELECT

       ! make sure we understand the BC kind

       SELECT CASE (bc_kind)
       CASE ("FLUX", "VELO", "ELEV", "ELEVELO")
       CASE ("UVEL", "VVEL")
       CASE ("CELL")
       CASE DEFAULT
          lerr = lerr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": ", TRIM(bc_kind), "?"
          CALL error_message(msg, fatal=.FALSE.)
       END SELECT


       IF (lerr .EQ. 0) THEN

          ! reread BC specific information from the line

          BACKSPACE(bcspec_iounit)

          SELECT CASE(bc_type)

          CASE("BLOCK")

             SELECT CASE(bc_extent)
             CASE("ALL")
                READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent,con_block
                block_bc(block)%bc_spec(num_bc)%con_block = con_block
                block_bc(block)%bc_spec(num_bc)%num_cell_pairs = 1
                block_bc(block)%bc_spec(num_bc)%start_cell(1) = 1
                block_bc(block)%bc_spec(num_bc)%end_cell(1) = maxidx
             CASE("PART")
                cells = -999
                READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent,con_block,cells(:)
                block_bc(block)%bc_spec(num_bc)%con_block = con_block
                CALL set_bc_part(bcspec_iounit, block_bc(block)%bc_spec(num_bc), &
                     &cells, maxidx, line, lerr)
                ! num_cells = COUNT(cells /= -999)
                ! num_cell_pairs = num_cells/2
                ! block_bc(block)%bc_spec(num_bc)%num_cell_pairs = num_cells/2
                ! block_bc(block)%bc_spec(num_bc)%start_cell(1:num_cell_pairs) = cells(1:num_cells:2)
                ! block_bc(block)%bc_spec(num_bc)%end_cell(1:num_cell_pairs) = cells(2:num_cells:2)
             CASE DEFAULT
                lerr = lerr + 1
                WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                     &": ", TRIM(bc_extent), "?"
                CALL error_message(msg, fatal=.FALSE.)
                READ (bcspec_iounit, *)
             END SELECT

          CASE("TABLE")
             SELECT CASE(bc_extent)
             CASE("ALL")
                READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent,file_name
                table_count = table_count + 1 ! keep a running count of the number of tables
                block_bc(block)%bc_spec(num_bc)%table_num = table_count
                table_bc(table_count)%file_name = file_name ! associate file and table number
                block_bc(block)%bc_spec(num_bc)%num_cell_pairs = 1
                block_bc(block)%bc_spec(num_bc)%start_cell(1) = 1
                block_bc(block)%bc_spec(num_bc)%end_cell(1) = maxidx
             CASE("PART")
                cells = -999
                READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent,file_name,cells(:)
                table_count = table_count + 1 ! keep a running count of the number of tables
                block_bc(block)%bc_spec(num_bc)%table_num = table_count
                table_bc(table_count)%file_name = file_name ! associate file and table number
                CALL set_bc_part(bcspec_iounit, block_bc(block)%bc_spec(num_bc), &
                     &cells, maxidx, line, lerr)
             CASE DEFAULT
                lerr = lerr + 1
                WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                     &": ", TRIM(bc_extent), "?"
                CALL error_message(msg, fatal=.FALSE.)
                READ (bcspec_iounit, *)
             END SELECT

             SELECT CASE (block_bc(block)%bc_spec(num_bc)%bc_kind)
             CASE ("FLUX")
                p = block_bc(block)%bc_spec(num_bc)%num_cell_pairs
                cmin = MINVAL(block_bc(block)%bc_spec(num_bc)%start_cell(1:p))+1
                cmax = MAXVAL(block_bc(block)%bc_spec(num_bc)%end_cell(1:p))+1
                ALLOCATE(block_bc(block)%bc_spec(num_bc)%flux_area(cmin:cmax))
             END SELECT

          CASE ("SOURCE","SINK")

             cells = -999

             READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent,file_name, cells
             table_count = table_count + 1 ! keep a running count of the number of tables
             block_bc(block)%bc_spec(num_bc)%table_num = table_count
             table_bc(table_count)%file_name = file_name ! associate file and table number

             SELECT CASE(bc_extent)
             CASE ("PART")
                CALL set_bc_area(block_bc(block)%bc_spec(num_bc), cells, &
                     &xmax(block) - 1, ymax(block) - 1, line, lerr)
             CASE ("ALL") 
                block_bc(block)%bc_spec(num_bc)%start_cell(1) = 1
                block_bc(block)%bc_spec(num_bc)%end_cell(1) = xmax(block) - 1
                block_bc(block)%bc_spec(num_bc)%start_cell(2) = 1
                block_bc(block)%bc_spec(num_bc)%end_cell(2) = ymax(block) - 1
             CASE DEFAULT
                lerr = lerr + 1
                WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                     &": ", TRIM(bc_extent), "?"
                CALL error_message(msg, fatal=.FALSE.)
                READ (bcspec_iounit, *)
             END SELECT

          CASE ("WALL")
             cells = -999
             READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,junk1,cells(:)

             ! the first number indicates the row
             ! (uvel) or column (vvel) cell 

             block_bc(block)%bc_spec(num_bc)%con_block = junk1

             ! the remaining numbers indicate the
             ! range of cells over which the wall
             ! is to be placed.

             SELECT CASE (bc_kind)
             CASE ("UVEL")
                maxidx = ymax(block) - 1
                IF (junk1 .GT. xmax(block) - 1) THEN
                   lerr = lerr + 1
                   WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                        &": x (up/down) index out of range: ", junk1
                   CALL error_message(msg, fatal=.FALSE.)
                END IF
             CASE ("VVEL") 
                maxidx = xmax(block) - 1
                IF (junk1 .GT. ymax(block) - 1) THEN
                   lerr = lerr + 1
                   WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                        &": y (left/right) index out of range: ", junk1
                   CALL error_message(msg, fatal=.FALSE.)
                END IF
             CASE DEFAULT
                lerr = lerr + 1
                WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                     &": this should not happen (fix me!)"
                CALL error_message(msg, fatal=.FALSE.)
             END SELECT
             CALL set_bc_part(bcspec_iounit, block_bc(block)%bc_spec(num_bc), &
                  &cells, maxidx, line, lerr)
          CASE ("DEAD")

             ! whatever is in bc_kind is not used

             cells = -999
             READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,cells(:)

             !
             ! the dead zone is specified as a
             ! rectangle of cells: imin, jmin,
             ! imax, jmax (like src/sink)

             CALL set_bc_area(block_bc(block)%bc_spec(num_bc), cells, &
                  &xmax(block) - 1, ymax(block) - 1, line, lerr)

           do_rptdead = .TRUE.

          CASE ("ZEROG")

             ! whatever is in bc_kind is ignored

             SELECT CASE(bc_extent)
             CASE("ALL")
                READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent
                block_bc(block)%bc_spec(num_bc)%num_cell_pairs = 1
                block_bc(block)%bc_spec(num_bc)%start_cell(1) = 1
                block_bc(block)%bc_spec(num_bc)%end_cell(1) = maxidx
             CASE("PART")
                cells = -999
                READ(bcspec_iounit,*)block,bc_loc,bc_type,bc_kind,bc_extent,cells(:)
                CALL set_bc_part(bcspec_iounit, block_bc(block)%bc_spec(num_bc), &
                     &cells, maxidx, line, lerr)
             CASE DEFAULT
                lerr = lerr + 1
                WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                     &": ", TRIM(bc_extent), "?"
                CALL error_message(msg, fatal=.FALSE.)
                READ (bcspec_iounit, *)
             END SELECT
          CASE DEFAULT
             lerr = lerr + 1
             WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
                  &": ", TRIM(bc_type), "?"
             CALL error_message(msg, fatal=.FALSE.)
             READ (bcspec_iounit, *)
          END SELECT

       END IF
       ierr = ierr + lerr
    END DO



200 CLOSE(bcspec_iounit)

    IF (ierr .GT. 0) THEN
       WRITE(msg, *) TRIM(bcspecs_name), ": ", ierr, " errors"
       CALL error_message(msg, fatal=.TRUE.)
    END IF
    CALL status_message("done reading bc specifications file")

  END SUBROUTINE read_bcspecs

  ! ----------------------------------------------------------------
  ! SUBROUTINE set_bc_part
  ! ----------------------------------------------------------------
  SUBROUTINE set_bc_part(iounit, spec, cells, maxidx, line, ierr)

    IMPLICIT NONE
    INTEGER, INTENT(IN) :: iounit, cells(:), maxidx, line
    TYPE (bc_spec_struct), INTENT(INOUT) :: spec
    INTEGER, INTENT(INOUT) :: ierr

    INTEGER :: num_cells, num_cell_pairs, i
    CHARACTER (LEN=1024) :: msg

    num_cells = COUNT(cells .NE. -999)
    num_cell_pairs = num_cells/2
    spec%num_cell_pairs = num_cell_pairs
    spec%start_cell(1:num_cell_pairs) = cells(1:num_cells:2)
    spec%end_cell(1:num_cell_pairs) = cells(2:num_cells:2)

    ! check the cell numbers

    DO i = 1, spec%num_cell_pairs
       IF (spec%start_cell(i) .LT. 1 .OR. spec%start_cell(i) .GT. maxidx) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": start index ", spec%start_cell(i), ' out of range'
          CALL error_message(msg, fatal=.FALSE.)
       END IF
       IF (spec%end_cell(i) .LT. 1 .OR. spec%end_cell(i) .GT. maxidx) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": end index ", spec%end_cell(i), ' out of range'
          CALL error_message(msg, fatal=.FALSE.)
       END IF
    END DO


  END SUBROUTINE set_bc_part

  ! ----------------------------------------------------------------
  ! SUBROUTINE set_bc_area
  ! ----------------------------------------------------------------
  SUBROUTINE set_bc_area(spec, cells, xmax, ymax, line, ierr)

    IMPLICIT NONE

    INTEGER, INTENT(IN) :: cells(:), xmax, ymax, line
    TYPE (bc_spec_struct), INTENT(INOUT) :: spec
    INTEGER, INTENT(INOUT) :: ierr

    INTEGER :: num_cells, num_cell_pairs, i
    CHARACTER (LEN=1024) :: msg

    num_cells = COUNT(cells /= -999)

    IF (num_cells .NE. 4) THEN
       ierr = ierr + 1
       WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
            &": four indices expected, got ", num_cells
       CALL error_message(msg, fatal=.FALSE.)
    ELSE
       num_cell_pairs = num_cells/2
       spec%num_cell_pairs = num_cells/2
       spec%start_cell(1:num_cell_pairs) = cells(1:num_cells:2)
       spec%end_cell(1:num_cell_pairs) = cells(2:num_cells:2)

       IF (spec%start_cell(1) .LT. 1 .OR. &
            &spec%start_cell(1) .GT. xmax) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": minimum x index out of range: ", spec%start_cell(1)
          CALL error_message(msg, fatal=.FALSE.)
       END IF

       IF (spec%end_cell(1) .LT. 1 .OR. &
            &spec%end_cell(1) .GT. xmax) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": maximum x index out of range: ", spec%end_cell(1)
          CALL error_message(msg, fatal=.FALSE.)
       END IF

       IF (spec%end_cell(1) .LT. spec%start_cell(1)) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": invalid x index range: ", spec%start_cell(1), " to ", &
               &spec%end_cell(1)
          CALL error_message(msg, fatal=.FALSE.)
       END IF

       IF (spec%start_cell(2) .LT. 1 .OR. &
            &spec%start_cell(2) .GT. ymax) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": minimum y index out of range: ", spec%start_cell(2)
          CALL error_message(msg, fatal=.FALSE.)
       END IF

       IF (spec%end_cell(2) .LT. 1 .OR. &
            &spec%end_cell(2) .GT. ymax) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": maximum y index out of range: ", spec%end_cell(2)
          CALL error_message(msg, fatal=.FALSE.)
       END IF

       IF (spec%end_cell(2) .LT. spec%start_cell(2)) THEN
          ierr = ierr + 1
          WRITE(msg, *) TRIM(bcspecs_name), ": error: line ", line, &
               &": invalid y index range: ", spec%start_cell(2), " to ", &
               &spec%end_cell(2)
          CALL error_message(msg, fatal=.FALSE.)
       END IF

    END IF

  END SUBROUTINE set_bc_area


END MODULE hydro_bc
